subsys/bluetooth/lib/ead.c

```diff
@@ -0,0 +1,119 @@
+/* Copyright (c) 2023 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/bluetooth/ead.h>
+
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+
+#include <zephyr/logging/log.h>
+#include <zephyr/bluetooth/crypto.h>
+
+/* This value is used to set the directionBit of the CCM  nonce to the MSB of the Randomizer field
+ * (see Supplement to the Bluetooth Core   Specification v11, Part A 1.23.3)
+ */
+#define BT_EAD_RANDOMIZER_DIRECTION_BIT 7
+
+/* Fixed value used for the Additional Authenticated Data (see Supplement to the Bluetooth Core
+ * Specification v11, Part A 1.23.3)
+ */
+static uint8_t bt_ead_aad[] = {0xEA};
+BUILD_ASSERT(sizeof(bt_ead_aad) == BT_EAD_AAD_SIZE);
+
+LOG_MODULE_REGISTER(bt_encrypted_ad_data, CONFIG_BT_EAD_LOG_LEVEL);
+
+static int bt_ead_generate_randomizer(uint8_t *randomizer)
+{
+	int err;
+
+	err = bt_rand(randomizer, BT_EAD_RANDOMIZER_SIZE);
+
+	if (err != 0) {
+		return err;
+	}
+
+	/* From Supplement to the Bluetooth Core Specification v11, Part A 1.23.3: The directionBit
+	 * of the CCM nonce shall be set to the most significant bit of the Randomizer field.
+	 */
+	randomizer[4] |= 1 << BT_EAD_RANDOMIZER_DIRECTION_BIT;
+
+	return 0;
+}
+
+static int bt_ead_generate_nonce(const uint8_t *iv, const uint8_t *randomizer, uint8_t *nonce)
+{
+	int err;
+	uint8_t new_randomizer[BT_EAD_RANDOMIZER_SIZE];
+
+	if (randomizer == NULL) {
+		err = bt_ead_generate_randomizer(new_randomizer);
+
+		if (err != 0) {
+			LOG_ERR("Failed to generate Randomizer");
+			return err;
+		}
+
+		randomizer = new_randomizer;
+	}
+
+	memcpy(&nonce[0], randomizer, BT_EAD_RANDOMIZER_SIZE);
+	memcpy(&nonce[BT_EAD_RANDOMIZER_SIZE], iv, BT_EAD_IV_SIZE);
+
+	return 0;
+}
+
+int bt_ead_encrypt(const uint8_t *session_key, const uint8_t *iv, const uint8_t *payload,
+		   int payload_size, uint8_t *encrypted_ad_data)
```

Comment:
Please add argument validation (i.e. NULL checks, and `payload_size` probably can't be `0` or larger than 254).